### DISCUSSION TITLE
How to revert a file, as :e! is bind to "Open File"?

### DISCUSSION THREAD

**Discussion by alanjds** [vscode-neovim/vscode-neovim#887] [General]
Some compatibility choices were made, but how to reset a file to the original (in-disk) contents, as `:e!` is bind to VSCode "open file"?

**Comment by asvetliakov** [comment]
`nnoremap <yourkey> :<C-u> call VSCodeNotify('workbench.action.files.revert')<CR>`

or

```
command! Revert call VSCodeNotify('workbench.action.files.revert')
AlterCommand r[evert] Revert
:r
```

Similar could be used to modify `e[dit]` command by your wish, see [this line](https://github.com/asvetliakov/vscode-neovim/blob/8cf96ab5eab8800eb98e11762edfd49da1c2ff16/vim/vscode-file-commands.vim#L29)

something like `1000u` or `:u0` may work too - but resets through vim undo history

also invoking vscode palette, typing "revert" and choosing `File: Revert File` command will work too

**Comment by justinmk** [comment]
@asvetliakov it is idiomatic to handle this via `BufEnter` (that's how directory plugins like netrw and  [dirvish](https://github.com/justinmk/vim-dirvish)) work.
 
So this could be handled by default, and will handle all kinds of commands besides `:edit`. When setting up vscode-neovim's buffer, add a buffer-local autocmd (i.e. `<buffer>`):

    autocmd BufEnter <buffer> call VSCodeNotify('workbench.action.files.revert')

**Comment by asvetliakov** [comment]
@justinmk I believe original issue was asking about just command `:e!` since it's currently mapped to 'workbench.action.files.openFile` (my bad actually). You don't need to revert buffer on enter with this extension since it's being initialized/wipeout-ed every time when you open/close the file in vscode.

**Comment by justinmk** [comment]
> You don't need to revert buffer on enter with this extension since it's being initialized/wipeout-ed every time when you open/close the file in vscode.

The reason I want to use `:edit!` is that sometimes I edit the file outside of VScode and then VScode does not autoload the latest version (for whatever reason, maybe it thinks I edited the file and there are conflicts, doesn't matter). I want to reload vscode's concept of the file, not vscode-neovim's (really long name to type out over and over, sigh) buffer.

> Similar could be used to modify `e[dit]` command by your wish, see [this line](https://github.com/asvetliakov/vscode-neovim/blob/8cf96ab5eab8800eb98e11762edfd49da1c2ff16/vim/vscode-file-commands.vim#L29)

I really hope we can eliminate  those commands. `AlterCommand` is pretty gross. Instead since vscode-neovim manages its own buffers, it is idiomatic for it to handle various things via buffer-local (`<buffer>`) autocmds.

For example instead of papering over `:write` and its variants, one could handle `FileWriteCmd`. 
 Though this will require creating a stub file since Vim doesn't support this for scratch buffers--maybe Nvim should introduce `BufWritePre` event so that `:write` doesn't complain in a scratch buffer.

**Comment by asvetliakov** [comment]
Neovim doesn't and can't handle any files, except few rare exceptions (help files). We just can't delegate file management to neovim due to big conceptual differences in handling files/documents between vscode & neovim. Ideally there should be `ext_files` extension or so which should ask external provider to load the file content (and perform other file/directory operations, e.g. there should be abstraction over the file reading/writing) to eliminate all these things.

>For example instead of papering over :write and its variants, one could handle FileWriteCmd.
Though this will require creating a stub file since Vim doesn't support this for scratch buffers--maybe Nvim should introduce BufWritePre event so that :write doesn't complain in a scratch buffer.

This won't work for external non-disk vscode files (i.e. remote/docker/wsl/non-file vscode document, such as git diff files , etc...). And vscode remote feature is one of the vscode big selling points, so i don't ready to sacrifice it. Need complete file abstraction layer for such cases

**Comment by justinmk** [comment]
I don't understand your last reply. I'm aware that vscode-neovim only maintains scratch buffers in Nvim, and that was the context of my previous reply.

**Comment by asvetliakov** [comment]
No-file in disk to reload (in case you're editing remote files in vscode) - `:edit!` won't work anyway (since there is nothing to reload). For `:write` something like `BufWritePre` may work, but i don't see any value to workaround one particular command but leave a bunch of anothers.

**Comment by justinmk** [comment]
> No-file in disk to reload

? That's up to `workbench.action.files.revert`. The point is just to invoke `workbench.action.files.revert` on `BufEnter` (which is triggered by `:edit` and equivalent commands).

**Comment by asvetliakov** [comment]
>. The point is just to invoke workbench.action.files.revert on BufEnter (which is triggered by :edit and equivalent commands).

If we'll remove `AlterCommand e[dit] Edit` you won't be able to do `:edit /etc/sysctl.conf` or `:edit ../blah.txt` (doesn't work now though) in case if the file is located somewhere remotely

For file reverting, consider the following case:
1) You did workspace edit in vscode (rename var operation for example) which produces changes in 5 files - this loads a files with changes in vscode but doesn't save them
2) You open (make active) a changed file in vscode -> making the file active (visible) creates neovim buffer
3) `BufEnter` is invoked and file will be reverted which obviously you don't want

Also due to vscode api limitations it's not possible to get opened but non-active files - so it's not possible to load them early in neovim before the rename operation (to avoid reverting rename operation changes)

**Comment by justinmk** [comment]
> you won't be able to do `:edit /etc/sysctl.conf` or `:edit ../blah.txt` ... remotely

You mean, because netrw handles it? Or what?

> 3\. `BufEnter` is invoked and file will be reverted which obviously you don't want

Why would it be invoked? If something happens in vscode, then vscode-neovim knows this before doing anything in Nvim, and can teardown/rebuild the buffer or  use `:noautocmd` perhaps.

**Comment by asvetliakov** [comment]
TLDR: there is no possibility to integrate both neovim file commands & vscode (unless neovim provides file-abstraction layer)

>You mean, because netrw handles it? Or what?

`netrw` or any real files in neovim is no way:
Opening a real file in neovim may (and i'm sure will) produce many issues when either vscode or neovim will complain about external changes in files, since we need to consider that changes and especially save operation could be contributed by external plugins/vscode actions itself, not the user. For `netrw` forcing users to configure SSH to be able to edit WSL or files inside docker-container in vscode is no way. Not to mention that `netrw` won't give any useful host information which we can use in vscode to load the actual file (e.g. for WSL we need to construct `wsl+arch://etc/sysconf.ctl` URI to open a file in vscode, `netrw` won't give this info and manually looking through hosts & translating ip is error-prone and clunky)

Example 1:
You're on windows (and using windows version of neovim) and editing in WSL: `/home/user/a.ts` file. 
You're trying to run: `:edit ./b.ts` file?

To use `netrw` you must know IP which doesn't exist in WSL1 and dynamic in WSL2

Example 2:
You're on windows (and using windows version of neovim) and opened file in docker container in vscode, i.e. file with URI: `docker+mycontainerinstance://home/user/a.ts`.
You're trying to run `:edit ./b.ts`

You can't use `netrw` since this container instancec doesn't have sshd/ftpd installed. Even if it has we'll lose the context in vscode about that file is located in docker container, thus vscode remote extension host won't work correctly

>If something happens in vscode, then vscode-neovim knows this before doing anything in Nvim, and can teardown/rebuild the buffer or use :noautocmd perhaps.

There is no way to catch vscode operation before it actually happens. When you do variable rename - all that you'll get - is the document change event. There is no way to know if a some file was opened (but no visible) additionally during the operation (well, there is document opened event but it's not reliable source of truth because vscode opens various documents in the background as result of some operations)

>Why would it be invoked? 

I'll explain by example:
1) we have loaded in vscode 3 files (in tabs) in single editor column: 
`a.ts [visible], b.ts, c.ts`

in contrary, neovim has only one buffer for `a.ts` because `b.ts` and `c.ts` are not visible for vscode extensions (current vscode API can only tell you about `a.ts`)

2) Now we switch active tab to `b.ts` - we destroy neovim buffer for `a.ts` and creating new buffer for `b.ts`. Neovim now has only single buffer for `b.ts`, vscode has `a.ts, b.ts [visible], c.ts`

3) Now we rename something (func, var, etc) in `b.ts` which will affect both `a.ts` or `c.ts` too
Due to rename operation - both `a.ts` and `c.ts` are modified and contain non saved changes

4)  We switch to `a.ts` again - and once again, a new buffer for `a.ts` will be created. Since it's new `BufAdd`/`BufEnter` autocmd will be invoked and will revert changes you did in 3)

Even more, i believe (correct me if i'm wrong) `BufEnter` will revert your changes even without creating a new buffer - after any `nvim_win_set_current_buf` call, and we're calling it every time when active file is changed in vscode.

**Comment by justinmk** [comment]
> Opening a real file in neovim may (and i'm sure will) produce many issues when either vscode or neovim will complain about external changes in files,

I never suggested that. My suggestion was to setup event handlers, which then could immediately _close_ the file and pass the filename to vscode.

>  For netrw forcing users to configure SSH

I don't care about netrw. I mentioned it because I'm trying to understand why you mentioned remote files regarding `:edit`.

> There is no way to catch vscode operation before it actually happens.

Surely the plugin knows when a file is opened. But I'll take your word for it.

> unless neovim provides file-abstraction layer

We could certainly think about that, but it will help to have a some concrete requirements. My first thought is that it might be useful for plugins like vscode-neovim if they could override commands like `:edit`, i.e. (imaginary Lua API): 

    vim.commands.set('edit', function()
      vscode.foo()
    end)

**Comment by asvetliakov** [comment]
>My suggestion was to setup event handlers, which then could immediately close the file and pass the filename to vscode.

See my concern about unexpectedly reverting changes when using `BufEnter`

>I don't care about netrw. I mentioned it because I'm trying to understand why you mentioned remote files regarding :edit.

Currently `edit` is remapped to custom `Edit` which can allow to support both `edit!` and `edit <path>` in reliable way. If we'll remove `edit -> Edit` remapping to support `edit!` through buffer events, then `edit <path>` won't work anymore, at least for non-local (i.e. remote) files.

>We could certainly think about that, but it will help to have a some concrete requirements. My first thought is that it might be useful for plugins like vscode-neovim if they could override commands like :edit, i.e. (imaginary Lua API):

I'd suggest something like clipboard api and on file level, like:

```
vim.readFile((path, cwd) -> vscode.readFile(path, cwd))
vim.writeFile((path, cwd, bufId, contents, force) -> vscode.writeFile(path, cwd, bufId, contents, force))
vim.stat((path) -> ....);
...
```

overriding default commands natively (may be good idea only to allow it in remote ui context) will be also nice for other things.

**Comment by justinmk** [comment]
> I'd suggest something like clipboard api and on file level, like:
> 
> ```
> vim.readFile((path, cwd) -> vscode.readFile(path, cwd))
> vim.writeFile((path, cwd, bufId, contents, force) -> vscode.writeFile(path, cwd, bufId, contents, force))
> vim.stat((path) -> ....);
> ```

@asvetliakov `BufReadCmd` and `BufWriteCmd` events already exist for that purpose. 

Toy example just to demonstrate:

    :autocmd BufReadCmd * put =localtime()

That handler preempts all file loading, thus `:edit foo.txt` will be handled by it. And `:edit` (no args) could then do `workbench.action.files.revert`.

Similarly `BufWriteCmd` takes-over `:write` and friends. 

    :autocmd BufWriteCmd * call VscodeWriteFile()

Then `AlterCommand` hack is not needed.

Note: BufReadCmd, BufWriteCmd are triggered only for physical files. So to create a "mirror" buffer, use `:edit` with the actual filepath.

**Comment by danielb2** [comment]
using `:e!` to discard changes to the current buffer (reverting the file) doesn't work, but gives an error with:
```
workbench.action.files.openFile failed, args: [] error: command 'workbench.action.files.openFile' not found
```

However, the fix is rather simple:
```
if exists('g:vscode')
    " e! should reload the file in vscode
    command! -bang -nargs=? Edit call VSCodeNotify('workbench.action.files.revert'>
endif
```

It seems reasonable this should be the default. Is there any reason it's not?

**Comment by alanjds** [comment]
@asvetliakov said:
> I believe original issue was asking about just command :e! since it's currently mapped to 'workbench.action.files.openFile` (my bad actually). You don't need to revert buffer on enter with this extension since it's being initialized/wipeout-ed every time when you open/close the file in vscode.

Yes, this was the original issue. Am used to `:e!` to revert to in-disk version of an unsaved buffer

then @danielb2 said:
> However, the fix is rather simple:
> 
> ```
> if exists('g:vscode')
>     " e! should reload the file in vscode
>     command! -bang -nargs=? Edit call VSCodeNotify('workbench.action.files.revert'>
> endif
> ```
> 
> It seems reasonable this should be the default. Is there any reason it's not?

That fix is the one I am using for more than one year! Have been working very well.

Am +1 to promote it to be the default.

**Comment by theol0403** [comment]
I thought this change has been made in #557.
Is this not the case for you in the latest version?

**Comment by danielb2** [comment]
Confirmed working. It usually prompts me to restart, but it didn't and  I hadn't used it on this computer besides testing just now to see. Thank you

**Comment by alanjds** [comment]
> I thought this change has been made in #557.
> Is this not the case for you in the latest version?

Huh. Nice. Now I can take off this workaround from my .vimrc

Thanks!

**Comment by theol0403** [comment]
Moving to discussion as it contains lots of good information about improving nvim integration and removing altercommand fix.

**Reply by justinmk** [reply]
issues with good information belong in the issue tracker...

**Reply by theol0403** [reply]
Sorry, the issue was closed and the topic of the issue has been resolved. This is my first time managing a project. If you think it would make more sense as an issue I can try to move it back.
